examples/apps/ai-collab/src/components/TaskGroup.tsx

```diff
@@ -59,8 +68,91 @@ export function TaskGroup(props: {
 		newBranchTargetNode: SharedTreeTaskGroup;
 	}>();
 
+	const [undoStack, setUndoStack] = useState<Revertible[]>([]);
+	const [redoStack, setRedoStack] = useState<Revertible[]>([]);
+
 	useSharedTreeRerender({ sharedTreeNode: props.sharedTreeTaskGroup, logId: "TaskGroup" });
 
+	/**
+	 * Create undo and redo stacks of {@link Revertible}.
+	 */
+	useEffect(() => {
+		function onRevertibleDisposed(disposed: Revertible): void {
+			const redoIndex = redoStack.indexOf(disposed);
+			if (redoIndex === -1) {
+				const undoIndex = undoStack.indexOf(disposed);
+				if (undoIndex !== -1) {
+					setUndoStack((currUndoStack) => {
+						const newUndoStack = [...currUndoStack];
+						newUndoStack.splice(undoIndex, 1);
+						return newUndoStack;
+					});
+				}
+			} else {
+				setRedoStack((currRedostack) => {
+					const newRedoStack = [...currRedostack];
+					newRedoStack.splice(redoIndex, 1);
+					return newRedoStack;
+				});
+			}
+		}
+
+		/**
+		 * Instead of application developer manually managing the life cycle of the {@link Revertible} instances,
+		 * example app stores upto `MAX_STACK_SIZE` number of {@link Revertible} instances in the undo and redo stacks.
+		 * When the stack size exceeds `MAX_STACK_SIZE`, the oldest {@link Revertible} instance is disposed.
+		 * @param stack - The primary stack that the {@link Revertible} instance is being added to.
+		 * @param setstack - The setter function for the primary stack.
+		 */
+		function maintainStackSize(
```

Comment:
I find this function a bit weird. It takes (and calls) a React setter, but the way we're using it, it actually gets called recursively (potentially) inside that same setter. Once the stack is the correct size the recursive call won't happen anymore, but I think the general flow would be cleaner less bug-prone if this function just operated locally on the array, and returned a version of the array that is the right size (whether that is the original array if it was already below the max size, or a new one otherwise). The call to that could be done before we call any of the React setters, and then we can just call them in a single location with the final array that they need to contain. Thoughts?

Reply:
Thinking about it, I think that's a good point. I don't think I need to pass in the setter within the setter. Instead, I can manipulate the array and then use the react setter.